### DISCUSSION TITLE
Best way to implement a manually triggered task

### DISCUSSION THREAD

**Discussion by skid** [vendurehq/vendure#1090] [Q&A] [answered]
Not sure if this is more of a Nest.js question than Vendure, but here goes.

I want to implement a procedure that will sync some data with a remote server and I need it to be manually triggerable, as well as scheduled to execute periodically. What is the best way to implement the manual trigger? For example, can I do the following:

1. Implement the sync logic and IO in a service
2. Create a NestJS controller with a GET method that can trigger the action and call the service
3. Can I decorate a controller method with `@Allow(Permission.SuperAdmin)` (docs says it's for resolvers)
4. If so, what's the best practice for testing this method with a REST API client (can you do Bearer Token auth for REST methods?

Does this sound sensible, or are there better ways?

While here - how do I run a task periodically? Sorry if this is documented already :)

**Comment by michaelbromley** [comment, marked as answer]
Yes, you can apply the `Allow` decorator to a Controller method. I just updated the doc block to state this.

Auth is handled in the exact same way for REST as with GraphQL - either bearer token or cookie depending on your config. You can see the e2e tests that verify this here: https://github.com/vendure-ecommerce/vendure/blob/97edcb9a2998ad8e13e091064c19b827f9f56f36/packages/core/e2e/plugin.e2e-spec.ts#L116-L129

> how do I run a task periodically?

Take a look at the official [NestJS cron package](https://docs.nestjs.com/techniques/task-scheduling) - since Vendure plugins are essentially just NestJS modules, you should be able to use this exactly as it shows in the docs.
